libsolidity/formal/CHC.cpp

```diff
@@ -323,6 +376,109 @@ void CHC::unknownFunctionCall(FunctionCall const&)
+ * f_j, the function body block after the loop.
+ * Create the following edges:
+ * f_i -> loop_header
+ * loop_header -> loop_body
+ * loop_header -> f_j
+ * loop_body -> loop_header, if there were no nested loops inside this loop body.
+ * f_k -> loop_header, if there was at least one nested loop inside this loop body,
+ * where f_k is the function body block after the latest nested loop.
+ */
+void CHC::visitLoop(
+	BreakableStatement const& _loop,
+	Expression const* _condition,
+	Statement const& _body,
+	ASTNode const* _postLoop
+)
+{
+	solAssert(m_currentFunction, "");
+	auto const& functionBody = m_currentFunction->body();
+
+	// Create loop header block and edge from function body to loop header.
+	m_predicates[&_loop] = createBlock(sort(functionBody), "loop_" + to_string(_loop.id()));
+	smt::Expression loopHeader = predicateBodyCurrent(&_loop);
+	smt::Expression functionLoop = smt::Expression::implies(
+		m_path.back() && m_context.assertions(),
+		loopHeader
+	);
+	addRule(functionLoop, &functionBody, &_loop);
+
+	// Loop header block visits the condition and has out-edges
+	// to the loop body and back to the rest of the function.
+	pushBlock(loopHeader);
+
+	if (_condition)
+		_condition->accept(*this);
+	auto condition = _condition ? expr(*_condition) : smt::Expression(true);
+
+	// Create loop body entry block.
+	m_predicates[&_body] = createBlock(sort(functionBody), "loop_body_" + to_string(_body.id()));
+	// This predicate needs to be created here to take potential
+	// loop condition side-effects into account.
+	smt::Expression loopBody = predicateBodyCurrent(&_body);
+	smt::Expression bodyEdge = smt::Expression::implies(
+		loopHeader && m_context.assertions() && condition,
+		loopBody
+	);
+	addRule(bodyEdge, &_loop, &_body);
+
+	// Loop body block visit.
+	pushBlock(loopBody);
+
+	auto functionBlocks = m_functionBlocks;
+	_body.accept(*this);
+	// TODO handle properly when break/continue are supported.
+	if (_postLoop)
+		_postLoop->accept(*this);
+
+	// If there are nested inner loops new function blocks
+	// are created within this loop body.
+	// In that case the back edge needs to come from the function block.
+	auto fromPred = m_functionBlocks > functionBlocks ?
+		predicateEntry(&functionBody) :
+		predicateEntry(&_body);
+	smt::Expression backEdge = smt::Expression::implies(
+		fromPred && m_context.assertions(),
+		predicateBodyCurrent(&_loop)
+	);
+	addRule(backEdge, &_body, &_loop);
+
+	// Pop all function blocks created by nested inner loops
+	// to adjust the assertion context.
+	for (unsigned i = m_functionBlocks; i > functionBlocks; --i)
+		popBlock();
+	m_functionBlocks = functionBlocks;
+
+	// Create a new function block here such that the function
+	// index increases for outer loops.
+	// The predicate needs to be created here when the loop body
+	// predicate is on the top of the stack.
+	createFunctionBlock(functionBody);
+	smt::Expression functionAfter = predicateEntry(&functionBody);
```

Comment:
Where is the loop connected to this? Does this happen implicitly somewhere? I only see the "header->functionAfter" edge below, but not the "postLoop->functionAfter" edge... It's probably being added somewhere, but it's quite difficult to keep track - I wonder whether this whole business could be made a bit clearer by some refactoring and some more member variables...